src/mongocxx/test/collection.cpp

```diff
@@ -2808,4 +2812,141 @@ TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[
 }
 
+TEST_CASE("expose writeErrors[].errInfo", "[collection]") {
+
+    // A helper for checking that an error document is well-formed according to our requirements:
+    auto writeErrors_well_formed = [](const bsoncxx::document::view& reply_view) -> bool {
+        if (!reply_view["writeErrors"]) {
+            return false;
+        }
+
+        const auto& errdoc = reply_view["writeErrors"][0];
+
+        auto error_code = errdoc["code"].get_int32();
+
+        // The code should always be 121 (DocumentValidationFailure):
+        if (121 != error_code) {
+            std::ostringstream os;
+            os << "writeErrors expected to have code 121, but had " << error_code << " instead";
+            throw std::runtime_error(os.str());
+        }
+
+        // We require the "details" field be present:
+        if (!errdoc["errInfo"]["details"]) {
+            throw std::runtime_error("no \"details\" field in \"writeErrors\"");
+        }
+
+        return true;
+    };
+
+    // Set up our test environment:
+    instance::current();
+
+    auto client_opts = mongocxx::options::client();
+
+    // Listen to the insertion-failed event: we want to get a copy of the server's
+    // response so that we can compare it to the thrown exception later:
+    apm_opts.on_command_succeeded(
+        [&writeErrors_well_formed](const mongocxx::events::command_succeeded_event& ev) {
+            if(!writeErrors_well_formed(ev.reply()))
+        });
+
+    client_opts.apm_opts(apm_opts);
+
+    auto mongodb_client = mongocxx::client(uri{}, client_opts);
+
+    database db = mongodb_client["prose_test_expose_details"];
+
+    const std::string collname{"mongo_cxx_driver-expose_details"};
+
+    // Drop the existing collection, if any:
+    db[collname].drop();
+
+    // Make a new collection with validation checking:
+    collection coll = db.create_collection(
+        collname,
+        make_document(kvp("validator",
+                          make_document(kvp("field_x", make_document(kvp("$type", "string")))))));
+
+    SECTION("insert a valid document") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", "value0"));
+
+        REQUIRE_NOTHROW(coll.insert_one(entry.view()));
+    }
+
+    SECTION("cause a type violation on insert") {
+        bsoncxx::builder::basic::document entry;
+
+        entry.append(kvp("_id", bsoncxx::oid()), kvp("field_x", 42));
+
+        try {
+            coll.insert_one(entry.view());
+
+            // We should not make it here:
+            CHECK(false);
+        } catch (const operation_exception& e) {
+            auto rse = e.raw_server_error();
+
+            CHECK(rse.has_value());
+
+            CHECK(writeErrors_well_formed(*rse));
+        } catch (...) {
+            // An exception was thrown, but of the wrong type:
+            CHECK(false);
+        }
+    }
+
+    SECTION("can update valid object with a valid value type") {
```

Comment:
The only test case that is necessary to satisfy requirements is "cause a type violation on insert"

> Enable `command monitoring <../../command-monitoring/command-monitoring.rst>`_
to observe CommandSucceededEvents. Then, insert an invalid document (e.g.
``{x: 1}``) and assert that a WriteError occurs, that its code is ``121``
(i.e. DocumentValidationFailure), and that its ``details`` property is
accessible. Additionally, assert that a CommandSucceededEvent was observed and
that the ``writeErrors[0].errInfo`` field in the response document matches the
WriteError's ``details`` property.

Each of the write commands go through `bulk_op.execute()`. I suggest removing the extra tests to simplify.

Reply:
+1